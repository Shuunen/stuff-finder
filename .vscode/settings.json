{
  "cSpell.words": [
    "onnomatch",
    "onresult",
    "onspeechend"
  ],
  "cSpell.ignoreWords": [
    "camwiegert"
  ]
}