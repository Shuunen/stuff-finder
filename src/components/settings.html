<div class="settings--trigger" onclick="emit('settings-toggle')">
    <svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"
        stroke-linecap="round" stroke-linejoin="round" class="feather feather-settings">
        <circle cx="12" cy="12" r="3"></circle>
        <path
            d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z">
        </path>
    </svg>
</div>

<div class="settings backdrop hidden">
    <div class="settings--modal center">
        <h2>Settings</h2>
        <p class="settings--message error"></p>
        <div class="row wrap center">
            <label class="setting col center">
                <a href="//airtable.com/api" target="_blank" class="mb1">Airtable api base</a>
                <input name="airtable-api-base" onkeyup="emit('settings-validate')" placeholder="your api base" pattern="^app\w{14}$" required
                    maxlength="17" autofocus>
            </label>
            <label class="setting col center mt1">
                <a href="//airtable.com/account" target="_blank" class="mb1">Airtable api key</a>
                <input name="airtable-api-key" onkeyup="emit('settings-validate')" placeholder="your api key" pattern="^key\w{14}$" required
                    maxlength="17">
            </label>
            <label class="setting col center mt1">
                Airtable table name
                <input name="airtable-table-name" class="mts" onkeyup="emit('settings-validate')" pattern="^\S+$" placeholder="your table name"
                    required>
            </label>
            <label class="setting col center mt1">
                Airtable view name
                <input name="airtable-view-name" class="mts" onkeyup="emit('settings-validate')" pattern="^\S+$" placeholder="your table name"
                    required value="stuff-finder">
            </label>
        </div>
        <small class="mt2">
            Your credentials will be used to fetch/modify your own data stored on Airtable. These credentials are saved in your browser to avoid
            typing these over and over. Feel free to check this app source code on the <a href="https://github.com/Shuunen/stuff-finder"
                target="_blank">Github project page</a>.
        </small>
        <div class="row center mt3">
            <button onclick="emit('settings-toggle')">üóô Close</button>
            <button class="settings--save" onclick="emit('settings-save')">Save ‚úîÔ∏è</button>
        </div>
    </div>
</div>

<script>
    class Settings extends window.BaseModel {
      get isDisplayed () {
        return !this.settingsEl.classList.contains('hidden')
      }

      get formIsValid () {
        return (this.settings.base.checkValidity() && this.settings.key.checkValidity())
      }

      constructor () {
        super('settings')
        this.log('constructor')
        this.settings = {}
      }

      setupElements () {
        this.settingsEl = document.querySelector('.settings')
        this.settingsTriggerEl = document.querySelector('.settings--trigger')
        this.settingsSaveEl = document.querySelector('.settings--save')
        this.settingsMessageEl = document.querySelector('.settings--message')
        this.settings.base = document.querySelector('.settings--modal input[name="airtable-api-base"]')
        this.settings.key = document.querySelector('.settings--modal input[name="airtable-api-key"]')
        this.settings.table = document.querySelector('.settings--modal input[name="airtable-table-name"]')
        this.settings.view = document.querySelector('.settings--modal input[name="airtable-view-name"]')
      }

      setupListeners () {
        this.on('settings-set', this.setSettings)
        this.on('settings-action-required', this.setActionRequired)
        this.on('settings-validate', this.validate)
        this.on('settings-toggle', this.toggle)
        this.on('settings-save', this.save)
      }

      onLoad () {
        this.validate()
      }

      toggle () {
        this.validate()
        if (this.isDisplayed) {
          this.fadeOut(this.settingsEl)
        } else {
          this.settingsEl.classList.remove('hidden')
        }
      }

      setSettings (data) {
        Object.keys(data).forEach(key => {
          const value = data[key]
          if (this.settings[key]) {
            this.settings[key].value = value
          }
        })
      }

      getSettings () {
        const settings = {}
        Object.keys(this.settings).forEach(key => {
          settings[key] = this.settings[key].value
        })
        return settings
      }

      validate () {
        this.log('validate...')
        if (this.formIsValid) {
          this.settingsSaveEl.removeAttribute('disabled')
        } else {
          this.settingsSaveEl.setAttribute('disabled', true)
        }
      }

      save () {
        if (this.formIsValid) {
          this.emit('settings-set', this.getSettings())
          this.toggle()
        }
      }

      setActionRequired (data) {
        if (typeof data !== 'object') {
          this.error('settings-action-required is expecting an object')
        }
        const required = typeof data.required === 'boolean' ? data.required : false
        this.log('set action require to', required)
        this.settingsTriggerEl.classList.toggle('action-required', required)
        if (data.message) {
          this.log('set message to', data.message)
          this.settingsMessageEl.textContent = data.message
        }
      }
    }
    // eslint-disable-next-line no-new
    new Settings()
</script>

<style>
    .settings {
      z-index: var(--elevation-bear, 50);
    }

    .settings--trigger {
      color: var(--color-grey, grey);
      cursor: pointer;
      height: 2.5rem;
      overflow: hidden;
      position: absolute;
      right: 2rem;
      top: 2rem;
      width: 2.5rem;
      z-index: var(--elevation-child, 30);
    }

    .settings--trigger.action-required {
      animation-duration: 2s;
      animation-fill-mode: both;
      animation-iteration-count: infinite;
      animation-name: bounce;
      color: var(--color-primary, steelblue);
      transform-origin: center bottom;
    }

    .settings--modal {
      background-color: var(--color-white, whitesmoke);
      border: .3rem solid var(--color-primary, steelblue);
      border-radius: var(--border-radius, .3rem);
      max-height: 90%;
      max-width: 50rem;
      overflow: auto;
      padding: .7rem 2rem 1rem;
      width: var(--settings-width, 80vw);
      z-index: var(--elevation-giraffe, 100);
    }

    @media only screen and (max-width: 600px) {
      .settings--modal {
        border: none;
        height: 100%;
        max-height: inherit;
        max-width: inherit;
        padding: .7rem;
        width: 100%;
      }
    }

    @keyframes bounce {
      from,
      20%,
      53%,
      80%,
      to {
        animation-timing-function: cubic-bezier(.215, .61, .355, 1);
        transform: translate3d(0, 0, 0);
      }

      40%,
      43% {
        animation-timing-function: cubic-bezier(.755, .05, .855, .06);
        transform: translate3d(0, -30px, 0);
      }

      70% {
        animation-timing-function: cubic-bezier(.755, .05, .855, .06);
        transform: translate3d(0, -15px, 0);
      }

      90% {
        transform: translate3d(0, -4px, 0);
      }
    }
</style>
