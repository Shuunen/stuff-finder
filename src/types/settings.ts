export interface AppCredentials extends Record<string, unknown> {
  base: string
  key: string
  table: string
  view: string
  wrap: string
}
