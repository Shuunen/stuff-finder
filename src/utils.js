export const dom = (type, content = '', classes = '') => {
  const element = document.createElement(type)
  element.className = classes
  element.innerHTML = content
  return element
}

export const div = (classes, content = '') => dom('div', content, classes)

export const img = (alt, source, classes = '') => {
  const element = dom('img', '', classes)
  element.alt = alt
  element.src = source
  return element
}

export const button = (content, classes = '') => {
  return dom('button', content, `bg-blue-800 m-auto sm:ml-0 px-4 py-1 ${classes}`)
}

export const p = (content, classes = '') => dom('p', content, classes)
