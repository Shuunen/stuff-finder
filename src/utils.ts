import { dom, emit } from 'shuutils'

export const button = (content: string, classes = '', secondary = false): HTMLButtonElement => {
  const theme = secondary ? 'hover:opacity-100 opacity-80' : 'from-purple-500 to-purple-700 text-white hover:from-purple-700 hover:to-purple-900'
  const button_ = dom('button', `text-lg h-10 max-w-xs px-12 md:text-base bg-gradient-to-tr shadow-md hover:shadow-lg rounded transition-all duration-200 each-in-out sm:px-6 ${theme} ${classes}`, content)
  button_.type = 'button'
  return button_
}

export const showError = (message: string) => {
  console.error(message)
  emit('app-toaster--show', { type: 'error', message })
}

export const showLog = (message: string, data = '') => {
  console.log(message, data)
  emit('app-toaster--show', { type: 'info', message })
}
