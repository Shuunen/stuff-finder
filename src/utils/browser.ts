import { dom, emit, sleep, storage, tw } from 'shuutils'
import { JSON_HEADERS } from '../constants'
import type { AirtableRecord, AppToasterShowEvent } from '../types'
import { urlToUuid } from './url'

export const button = (content: string, classes = '', secondary = false): HTMLButtonElement => {
  const theme = secondary ? tw('opacity-80 hover:opacity-100') : tw('from-purple-500 to-purple-700 text-white hover:from-purple-700 hover:to-purple-900')
  const button_ = dom('button', tw(`h-10 max-w-xs rounded bg-gradient-to-tr px-12 text-lg shadow-md transition-all duration-200 ease-in-out hover:shadow-lg sm:px-6 md:text-base ${theme} ${classes}`), content)
  button_.type = 'button'
  return button_
}

/* eslint-disable no-console */
export const logger = {
  error: (message: string, ...data: readonly unknown[]): void => { console.error(message, ...data) },
  showError: (message: string, ...data: readonly unknown[]): void => {
    console.error(message, ...data)
    emit<AppToasterShowEvent>('app-toaster--show', { type: 'error', message })
  },
  log: (message: string, ...data: readonly unknown[]): void => { console.log(message, ...data) },
  showLog: (message: string, ...data: readonly unknown[]): void => {
    console.log(message, ...data)
    emit<AppToasterShowEvent>('app-toaster--show', { type: 'info', message })
  },
}
/* eslint-enable no-console */

const request = async <T> (method: 'get' | 'patch' | 'post', url: string, data?: Record<string, unknown>): Promise<T> => {
  const options: RequestInit = { headers: JSON_HEADERS, method }
  if (data) options.body = JSON.stringify(data)
  return fetch(url, options).then(async (response: Response) => response.json()).catch(error => { logger.showError((error as Error).message) }) as Promise<T>
}

export const patch = async (url: string, data: Record<string, unknown>): Promise<AirtableRecord> => request('patch', url, data)

export const post = async (url: string, data: Record<string, unknown>): Promise<AirtableRecord> => request('post', url, data)

export const get = async <T> (url: string): Promise<T> => {
  const uuid = urlToUuid(url)
  const cached = storage.get<T>(uuid, undefined, sessionStorage)
  if (cached !== undefined) return cached
  const response = await request<T>('get', url)
  storage.set(uuid, response, sessionStorage)
  return response
}

/* eslint-disable no-restricted-properties, unicorn/prefer-spread */
export const find = {
  one: <T extends Element = Element> (selector: string, context: Document | Element = document): T => {
    const element = context.querySelector<T>(selector)
    if (!element) throw new Error(`no element found for selector "${selector}"`)
    return element
  },
  oneOrNone: <T extends Element = Element> (selector: string, context: Document | Element = document): T | null => context.querySelector<T>(selector),
  all: <T extends Element = Element> (selector: string, context: Document | Element = document): T[] => {
    const elements = context.querySelectorAll<T>(selector)
    if (elements.length === 0) throw new Error(`no elements found for selector "${selector}"`)
    return Array.from(elements)
  },
  allOrNone: <T extends Element = Element> (selector: string, context: Document | Element = document): T[] => Array.from(context.querySelectorAll<T>(selector)),
}
/* eslint-enable no-restricted-properties, unicorn/prefer-spread */

export const fadeIn = async (element: HTMLElement): Promise<void> => {
  if (!element.classList.contains('app-hide')) { logger.error('please add "app-hide" class before mounting dom element and then call fade-in'); return }
  element.classList.remove('hidden')
  await sleep(10)
  element.style.opacity = '1'
}

export const fadeOut = async (element: HTMLElement, destroy = false): Promise<void> => {
  element.classList.add('app-hide')
  element.style.opacity = '0'
  await sleep(350)
  element.classList.add('hidden')
  if (!destroy) return
  await sleep(350)
  element.remove()
}

export const valuesToOptions = (array: readonly string[], selected?: string): string => {
  return array.map(value => `<option value="${value}" ${selected === value ? 'selected' : ''}>${value}</option>`).join('')
}

export const isVisible = (element: Element | HTMLElement | undefined): boolean => {
  if (!element) {
    logger.error('element is undefined, so it is not visible')
    return false
  }
  const { top, bottom, width, height } = element.getBoundingClientRect()
  const visible = width > 0 && height > 0 && top >= 0 && bottom <= window.innerHeight
  // logger.log(`element is ${visible ? '' : 'not '}visible`, element)
  return visible
}
