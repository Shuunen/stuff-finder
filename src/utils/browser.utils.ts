import { dom, sleep, storage, tw } from 'shuutils'
import { delays, jsonHeaders } from '../constants'
import { logger } from './logger.utils'
import { urlToUuid } from './url.utils'

// eslint-disable-next-line etc/no-misused-generics
async function request<ResponseType = unknown> (method: 'GET' | 'PATCH' | 'POST', url: string, data?: Record<string, unknown>) {
  const options: RequestInit = { headers: jsonHeaders, method }
  if (data) options.body = JSON.stringify(data)
  const response = await fetch(url, options).catch((error: unknown) => { logger.showError(error) })
  if (!response) throw new Error('no response')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return (await response.json()) as ResponseType
}

export function button (content: string, classes = '', isSecondary = false) {
  const theme = isSecondary ? tw('opacity-80 hover:opacity-100') : tw('from-purple-500 to-purple-700 text-white hover:from-purple-700 hover:to-purple-900')
  const element = dom('button', tw(`h-10 max-w-xs rounded bg-gradient-to-tr px-12 text-lg shadow-md transition-all duration-200 ease-in-out hover:shadow-lg sm:px-6 md:text-base ${theme} ${classes}`), content)
  element.type = 'button'
  return element
}

// eslint-disable-next-line etc/no-misused-generics
export async function patch<ResponseType> (url: string, data: Record<string, unknown>) {
  return await request<ResponseType>('PATCH', url, data)
}

// eslint-disable-next-line etc/no-misused-generics
export async function post<ResponseType> (url: string, data: Record<string, unknown>) {
  return await request<ResponseType>('POST', url, data)
}

// eslint-disable-next-line etc/no-misused-generics
export async function get<ResponseType> (url: string, willCacheResponse = true) {
  const uuid = urlToUuid(url)
  if (willCacheResponse) {
    const cached = storage.get<ResponseType>(uuid)
    if (cached !== undefined) return cached
  }
  const response = await request<ResponseType>('GET', url)
  if (willCacheResponse) storage.set(uuid, response)
  return response
}

/* eslint-disable etc/no-misused-generics */
export const find = {
  all: <Type extends Element = Element> (selector: string, context: Document | Element = document) => {
    const elements = context.querySelectorAll<Type>(selector)
    if (elements.length === 0) throw new Error(`no elements found for selector "${selector}"`)
    return Array.from(elements)
  },
  allOrNone: <Type extends Element = Element> (selector: string, context: Document | Element = document): Type[] => Array.from(context.querySelectorAll<Type>(selector)),
  one: <Type extends Element = Element> (selector: string, context: Document | Element = document) => {
    const element = context.querySelector<Type>(selector)
    if (!element) throw new Error(`no element found for selector "${selector}"`)
    return element
  },
  oneOrNone: <Type extends Element = Element> (selector: string, context: Document | Element = document): Type | null => context.querySelector<Type>(selector),
}
/* eslint-enable etc/no-misused-generics */

export async function fadeIn (element: HTMLElement) {
  if (!element.classList.contains('app-hide')) { logger.error('please add "app-hide" class before mounting dom element and then call fade-in'); return }
  element.classList.remove('hidden')
  await sleep(10) // eslint-disable-line @typescript-eslint/no-magic-numbers
  element.style.opacity = '1' // eslint-disable-line no-param-reassign
}

export async function fadeOut (element: HTMLElement, willDestroyAfter = false) {
  element.classList.add('app-hide')
  element.style.opacity = '0' // eslint-disable-line no-param-reassign
  await sleep(delays.large)
  element.classList.add('hidden')
  if (!willDestroyAfter) return
  await sleep(delays.large)
  element.remove()
}

export function valuesToOptions (array: readonly string[], selected?: string) {
  return array.map(value => `<option value="${value}" ${selected === value ? 'selected' : ''}>${value}</option>`).join('')
}

export function isVisible (element: Element | HTMLElement | undefined) {
  if (!element) {
    logger.error('element is undefined, so it is not visible')
    return false
  }
  const { height, width } = element.getBoundingClientRect()
  return width > 0 || height > 0
}
