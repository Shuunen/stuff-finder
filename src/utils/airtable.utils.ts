/* eslint-disable functional/immutable-data */
import { safeParse } from 'valibot'
import { emptyItem } from '../constants'
import { del, get, patch, post } from './browser.utils'
import { logger } from './logger.utils'
import { airtableDeleteResponseSchema, airtableMultipleResponseSchema, airtableSingleResponseSchema, type AirtableSingleRecordResponse, type Item, type ItemField, type ItemPhoto } from './parsers.utils'
import { state } from './state.utils'

const airtableBaseUrl = 'https://api.airtable.com/v0'

export const airtableMaxRequestPerSecond = 5

// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: <explanation>
export function getItemFieldsToPush (data: Item, currentState = state) { // eslint-disable-line max-statements, complexity, sonarjs/cognitive-complexity
  const fields: Partial<AirtableSingleRecordResponse['fields']> = {}
  if (data.barcode.length > 0) fields.barcode = data.barcode
  if (data.box.length > 0) fields.box = data.box
  if (data.brand.length > 0) fields.brand = data.brand
  if (data.category.length > 0) fields.category = data.category
  if (data.details.length > 0) fields.details = data.details
  if (data.drawer.length > 0) fields.drawer = data.drawer
  if (data.location.length > 0) fields.location = data.location
  if (data.name.length > 0) fields.name = data.name
  /* c8 ignore next 2 */
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (data.photo !== undefined && data.photo.length > 0 && fields.photo?.[0]?.url !== data.photo[0]?.url) fields.photo = [{ url: data.photo[0]?.url } as unknown as ItemPhoto] // we don't need the whole object
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  if (data.price !== undefined && data.price > -1) fields.price = data.price
  if (data.reference.length > 0) fields.reference = data.reference
  if (data.status.length > 0) fields.status = data.status
  fields['ref-printed'] = data['ref-printed']
  if (data.id.length > 0) {
    const existing = currentState.items.find(existingItem => existingItem.id === data.id)
    if (!existing) throw new Error('existing item not found locally')
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const dataFields = Object.keys(fields) as ItemField[]
    for (const field of dataFields) {
      /* c8 ignore next 2 */
      if (field === 'id') continue // eslint-disable-line no-continue
      const hasSamePhoto = (field === 'photo' && existing.photo && existing.photo[0]?.url === fields.photo?.[0]?.url) ?? false
      const hasSameValue = existing[field] === fields[field]
      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
      if (hasSamePhoto || hasSameValue) delete fields[field]
    }
  }
  return fields
}

export function airtableRecordToItem (record: AirtableSingleRecordResponse) {
  return {
    ...emptyItem,
    ...record.fields,
    id: record.id,
  }
}

export function isLocalAndRemoteSync (records: AirtableSingleRecordResponse[], currentState = state) {
  const record = records.at(0)
  if (!record) throw new Error('remoteFirst is undefined')
  const remoteFirst = airtableRecordToItem(record)
  const localFirst = currentState.items.at(0)
  const isFirst = localFirst === undefined ? false : remoteFirst.id === localFirst.id && remoteFirst['updated-on'] === localFirst['updated-on']
  if (isFirst) return true
  const localLast = currentState.items.at(-1) // eslint-disable-line @typescript-eslint/no-magic-numbers
  return localLast === undefined ? false : remoteFirst.id === localLast.id && remoteFirst['updated-on'] === localLast['updated-on']
}

export async function getOneItem (id: Item['id'], getMethod = get) {
  const { base, table } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  const result = safeParse(airtableSingleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return airtableRecordToItem(result.output)
}

export async function getAllItems (offset?: string, getMethod = get) {
  const offsetParameter = offset === undefined ? '' : `&offset=${offset}`
  const sortByUpdatedFirst = '&sort%5B0%5D%5Bfield%5D=updated-on&sort%5B0%5D%5Bdirection%5D=desc'
  const { base, table, view } = state.credentials
  const url = `${airtableBaseUrl}/${base}/${table}?view=${view}${offsetParameter}${sortByUpdatedFirst}`
  const result = safeParse(airtableMultipleResponseSchema, await getMethod(url))
  if (!result.success) {
    logger.error(result.issues)
    throw new Error(`failed to fetch item, issue(s) : ${result.issues.map(issue => issue.message).join(', ')}`)
  }
  return result.output
}

export async function deleteItemRemotely (id: Item['id'], currentState = state, delMethod = del) {
  const { base, table } = currentState.credentials
  const url = `${airtableBaseUrl}/${base}/${table}/${id}`
  return safeParse(airtableDeleteResponseSchema, await delMethod(url))
}

// eslint-disable-next-line @typescript-eslint/max-params
export async function pushItemRemotely (item: Item, currentState = state, postMethod = post, patchMethod = patch) {
  const fields = getItemFieldsToPush(item, currentState)
  if (Object.keys(fields).length === 0) {
    logger.showLog('no update to push')
    return { output: { fields, id: item.id }, success: false } // eslint-disable-line @typescript-eslint/naming-convention
  }
  const data = { fields }
  const { base, table } = currentState.credentials
  const baseUrl = `${airtableBaseUrl}/${base}/${table}`
  if (item.id === '') return safeParse(airtableSingleResponseSchema, await postMethod(baseUrl, data))
  const url = `${baseUrl}/${item.id}`
  return safeParse(airtableSingleResponseSchema, await patchMethod(url, data))
}
