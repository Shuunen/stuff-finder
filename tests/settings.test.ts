import { expect, it } from 'vitest' // eslint-disable-line @typescript-eslint/no-shadow
import { validateForm } from '../src/utils/forms.utils'
import { settingsForm } from '../src/utils/settings.utils'

it('validateForm A default settingsForm is not changed & not valid', () => {
  const { hasChanged, updatedForm } = validateForm(settingsForm)
  expect(updatedForm).toMatchSnapshot()
  expect(updatedForm.isValid).toBe(false)
  expect(updatedForm.errorMessage).toMatchInlineSnapshot('"Airtable base is required"')
  expect(updatedForm).toStrictEqual(settingsForm)
  expect(hasChanged).toBe(false)
})
