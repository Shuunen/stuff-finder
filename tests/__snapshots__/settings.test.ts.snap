// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validateForm A default settingsForm is not changed & not valid 1`] = `
{
  "errorMessage": "Airtable base is required",
  "fields": {
    "base": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable base",
      "link": "https://airtable.com/api",
      "order": 10,
      "regex": /\\^app\\\\w\\{14\\}\\$/u,
      "value": "",
    },
    "table": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable table",
      "order": 30,
      "regex": /\\^\\[\\\\w-\\]\\{3,100\\}\\$/u,
      "value": "",
    },
    "token": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable personal access token",
      "link": "https://airtable.com/create/tokens",
      "order": 20,
      "regex": /\\^pat\\[\\\\w\\.\\]\\{50,100\\}\\$/u,
      "value": "",
    },
    "view": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable view",
      "order": 40,
      "regex": /\\^\\[\\\\w-\\]\\{3,100\\}\\$/u,
      "value": "",
    },
    "wrap": {
      "isRequired": false,
      "isValid": true,
      "label": "Wrap api key",
      "link": "https://wrapapi.com/user/api",
      "order": 50,
      "regex": /\\^\\\\w\\{32\\}\\$/u,
      "value": "",
    },
  },
  "isTouched": false,
  "isValid": false,
}
`;

exports[`validateForm A default settingsForm is not changed but invalid 1`] = `
{
  "errorMessage": "Airtable base is required",
  "fields": {
    "base": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable base",
      "link": "https://airtable.com/api",
      "order": 10,
      "regex": /\\^app\\\\w\\{14\\}\\$/u,
      "value": "",
    },
    "table": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable table",
      "order": 30,
      "regex": /\\^\\[\\\\w-\\]\\{3,100\\}\\$/u,
      "value": "",
    },
    "token": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable personal access token",
      "link": "https://airtable.com/create/tokens",
      "order": 20,
      "regex": /\\^pat\\[\\\\w\\.\\]\\{50,100\\}\\$/u,
      "value": "",
    },
    "view": {
      "isRequired": true,
      "isValid": false,
      "label": "Airtable view",
      "order": 40,
      "regex": /\\^\\[\\\\w-\\]\\{3,100\\}\\$/u,
      "value": "",
    },
    "wrap": {
      "isRequired": false,
      "isValid": true,
      "label": "Wrap api key",
      "link": "https://wrapapi.com/user/api",
      "order": 50,
      "regex": /\\^\\\\w\\{32\\}\\$/u,
      "value": "",
    },
  },
  "isTouched": false,
  "isValid": false,
}
`;
